File: nav_pkg/src/wp_node.py
#!/usr/bin/env python3
#conding: utf-8
import rclpy
from std_msgs.msg import String
import time
import sys

waypoints = ["1", "2", "3", "4"]
index = 0
waiting = False
node = None
waypoint_pub = None


def nav_result_callback(msg):
    global node, waiting, index
    node.get_logger().info("Received Result: " + msg.data)
    index += 1
    waiting = False

    time.sleep(1.0)  # Sleep for a short time before sending the next waypoint

    if index < len(waypoints):
        msg = String()
        msg.data = waypoints[index]
        waypoint_pub.publish(msg)
        node.get_logger().info(f"Send waypoint name : {msg.data}")
    else:
        node.get_logger().info("All waypoints have been sent.")



def main(args=None):
    global node, waypoint_pub
    rclpy.init(args=None)

    print("Hello World, I am the wp node![Python]\n");

    node = rclpy.create_node('wp_node')
    waypoint_pub = node.create_publisher(String, "waterplus/navi_waypoint", 10)
    result_sub = node.create_subscription(String, "waterplus/navi_result", nav_result_callback, 10)

    time.sleep(1.0)  # Sleep for a short time to ensure subscriptions are set up

    msg = String()
    msg.data = waypoints[index]

    waypoint_pub.publish(msg)
    node.get_logger().info(f"Send waypoint name : {msg.data}")

    rclpy.spin(node)

if __name__ == '__main__':
    main()

File: nav_pkg/src/wp_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

std::shared_ptr<rclcpp::Node> node;

void NavResultCallback(const std_msgs::msg::String::SharedPtr msg){
    RCLCPP_WARN(node->get_logger(), "Received Result: %s", msg->data.c_str());
}


int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);

    printf("Hello World, I am the wp node!\n");

    node = rclcpp::Node::make_shared("wp_node");

    std::shared_ptr<rclcpp::Publisher<std_msgs::msg::String>> waypoint_pub =
        node->create_publisher<std_msgs::msg::String>("waterplus/navi_waypoint", 10);

    std::shared_ptr<rclcpp::Subscription<std_msgs::msg::String>> result_sub =
        node->create_subscription<std_msgs::msg::String>("waterplus/navi_result", 10, NavResultCallback);

    // sleep
    rclcpp::sleep_for(std::chrono::milliseconds(1000));

    std_msgs::msg::String msg;

    if(argc > 1)
        msg.data = argv[1];
    else
        msg.data = "1";

    waypoint_pub->publish(msg);

    RCLCPP_INFO(node->get_logger(), "Send waypoint name : %s",msg.data.c_str());
    rclcpp::spin(node);

    return 0;
}



